dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/TimeToLive.java

```diff
@@ -0,0 +1,154 @@
+        notNull( params.getLatestEndDate(), "Object params.getLatestEndDate() must not be null" );
+        notNull( systemSettingManager, "Object systemSettingManager must not be null" );
+        this.params = params;
+        this.systemSettingManager = systemSettingManager;
+    }
+
+    /**
+     * Calculates the time to live based on specific internal rules.
+     *
+     * @return the computed TTL value in SECONDS.
+     */
+    @Override
+    public long compute()
+    {
+        final Instant oldestDate = params.getEarliestStartDate().toInstant();
+        final Instant mostRecentDate = params.getLatestEndDate().toInstant();
+        final long daysDiff = DAYS.between( oldestDate, mostRecentDate );
+        return calculationFor( daysDiff, mostRecentDate );
+    }
+
+    /**
+     * Execute the internal rules in order to calculate a TTL for the given
+     * parameters.
+     *
+     * @param daysDiff the difference of days between the oldest and most recent
+     *        date in the period.
+     * @param mostRecentDate the most recent date of the period.
+     * @return a time to live value in SECONDS.
+     */
+    long calculationFor( final long daysDiff, final Instant mostRecentDate )
+    {
+        /*
+         * If the difference between the most recent date and NOW is 0 (zero) it means
+         * the current year, so we increment the multiplier by 1 (one) avoiding
+         * multiplying by 0 (zero).
+         */
+        final long ttlMultiplierOrOne = YEARS.between( ofInstant( mostRecentDate, UTC ), now() ) + 1;
+        final long ttlInSeconds;
+
+        if ( daysDiff <= WEEKLY.value() )
+        {
+            ttlInSeconds = preDefinedTtlValueForKey(ANALYTICS_CACHE_TIMEOUT_FACTOR_WEEKLY_PERIOD_IN_SECONDS);
+        }
+        else if ( daysDiff <= MONTHLY.value() )
+        {
+            ttlInSeconds = preDefinedTtlValueForKey(ANALYTICS_CACHE_TIMEOUT_FACTOR_MONTHLY_PERIOD_IN_SECONDS);
+        }
+        else if ( daysDiff <= QUARTERLY.value() )
+        {
+            ttlInSeconds = preDefinedTtlValueForKey(ANALYTICS_CACHE_TIMEOUT_FACTOR_QUARTERLY_PERIOD_IN_SECONDS);
+        }
+        else if ( daysDiff <= SIX_MONTHS.value() )
+        {
+            ttlInSeconds = preDefinedTtlValueForKey(ANALYTICS_CACHE_TIMEOUT_FACTOR_SIX_MONTHS_PERIOD_IN_SECONDS);
+        }
+        else
+        {
+            ttlInSeconds = preDefinedTtlValueForKey(ANALYTICS_CACHE_TIMEOUT_FACTOR_YEARLY_OR_OVER_PERIOD_IN_SECONDS);
+        }
+        return ttlInSeconds * ttlMultiplierOrOne;
+    }
+
+    private Long preDefinedTtlValueForKey( final SettingKey ttlSettingKey )
+    {
+        final Long ttlInSeconds = (Long) systemSettingManager.getSystemSetting( ttlSettingKey );
+        final boolean ttlNotNullAndPositive = ttlInSeconds != null && ttlInSeconds > 0;
+
+        return ttlNotNullAndPositive ? ttlInSeconds : DEFAULT_TO_30_SECONDS;
+    }
+
+    enum Periods
```

Comment:
Rename Periods to `PeriodTypes`, i.e. Weekly is a period type, not a period.

Reply:
Not used anymore. Code removed